### PR TITLE
Tell subprocesses the path to self in an env variable #3778

### PR DESCRIPTION
**Pull request by vojtechkral**
I'm just setting the env var on the process itself, letting subprocesses inherit that, as it's easier than setting for each subprocess individually. I'm not entirely sure this is the right spot though.

Also, I should probably document this somewhere - what would be the best place?

### CODE REVIEW

**Comment by alexcrichton** [comment]
Could this be done in a perhaps less invasive way than setting an env var? Could we just set it for child processes where it's relevant?

**Comment by vojtechkral** [comment]
Sure, provided I'll be able to find the right / relevant places. I'll update...

**Comment by vojtechkral** [comment]
Ok, so I restricted the env var setting to external subcommands. (I'm still not sure about #3744, I'll update for that one later...)

File: src/cargo/ops/cargo_test.rs [hunk 1]
```diff
@@ -91,6 +92,9 @@ fn run_unit_tests(options: &TestOptions,
         };
         let mut cmd = compilation.target_process(exe, pkg)?;
         cmd.args(test_args);
+        if let Some(arg0) = env::args_os().next() {
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_test.rs]: If you follow the definition of `target_process` I think you'll find the `fill_env` function which is likely where this should go.

**Comment by vojtechkral** [comment]
Yeah, it occured to me tests would be nice too, this is a bit tricky though, I'm gonna need an 'answering subprocess'. I have no idea what the testing environment requirements are, can I assume it is unixy and can execute basic shell scripts?

**Comment by alexcrichton** [comment]
Oh you can take a look at `tests/*.rs`, there should be tons of examples of child processes and such.

**Comment by vojtechkral** [comment]
Uhh, lemme fix that...

**Comment by alexcrichton** [comment]
Reading this again, I think that we may also want to store this somewhere in `Config`? Right now literally passing `argv[0]` is just passing whatever was passed on the command line, not necessarily an absolute path. Perhaps we could use `current_exe` to populate this, and cache that value in `Config`?

**Comment by vojtechkral** [comment]
Yeah, I've just rewritten it using `current_exe`. I'll look at `Config`...

**Comment by vojtechkral** [comment]
Does something like `cargo_exe: LazyCell<PathBuf>,` sound good to you?

**Comment by alexcrichton** [comment]
Yeah sounds fine by me

**Comment by matklad** [comment]
>Also, I should probably document this somewhere - what would be the best place?

http://doc.crates.io/environment-variables.html seems like a good option.

**Comment by alexcrichton** [comment]
r=me with some docs @matklad pointed out, thanks @vojtechkral!

**Comment by alexcrichton** [comment]
(oops)

File: src/cargo/lib.rs [hunk 2]
```diff
@@ -41,6 +41,8 @@ use term::color::{BLACK};
 
 pub use util::{CargoError, CargoResult, CliError, CliResult, human, Config, ChainError};
 
+pub const CARGO_ENV: &'static str = "CARGO";
```
Comment (matklad) [inline src/cargo/lib.rs]: I think something more verbose like `CARGO_EXECUTABLE` would be a tiny bit better, but we already use plain `RUSTC` and `RUSRDOC` variables for similar purposes, and in this regard just `CARGO` is indeed the best name here.

Comment (vojtechkral) [inline src/cargo/lib.rs]: I don't really have a strong feeling one way or another ([hehe](https://www.youtube.com/watch?v=ussCHoQttyQ)) but I guess, like you said, `CARGO` is good for consistency with `RUSTC` ...

**Comment by vojtechkral** [comment]
Writing the documentation makes me think whether maybe we should just call `fill_env` for external subcommands as well... Would that make sense @alexcrichton ?

**Comment by alexcrichton** [comment]
Yeah those two code paths are unfortunately pretty different

**Comment by vojtechkral** [comment]
(added doc + squashed)

**Comment by alexcrichton** [comment]
@bors: r+

Thanks @vojtechkral!

**Pull request merged**